## This notebook will explain about the basic architecture on small image data like MNIST 

### 1.Loading the MNIST data 

In [1]:
from keras.datasets import mnist

C:\Users\vipul\Anaconda3\lib\site-packages\h5py\__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [2]:
#train test split of the dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11493376/11490434 [==============================] - 25s 2us/step


In [3]:
train_images.shape

(60000, 28, 28)

* There are 60000 gray scale images of digits(0-9) in training dataset , each of dim 28x28

In [5]:
test_images.shape

(10000, 28, 28)

* There are 10000 gray scale images of digits(0-9) in test dataset , each of dim 28x28 .

### 2.Network archietecture

In [6]:
from keras import models
from keras import layers

In [26]:
neural_net=models.Sequential()

#layer 1-- hidden layer of 256 units
neural_net.add(layers.Dense(256,activation='relu',input_shape=(28*28,)))
#layer 2-- hidden layer of 128 units
neural_net.add(layers.Dense(128,activation='relu'))

#output layer
neural_net.add(layers.Dense(10,activation='softmax'))
#softmax because it is a multiclass classification

neural_net.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_10 (Dense)             (None, 256)               200960    
_________________________________________________________________
dense_11 (Dense)             (None, 128)               32896     
_________________________________________________________________
dense_12 (Dense)             (None, 10)                1290      
Total params: 235,146
Trainable params: 235,146
Non-trainable params: 0
_________________________________________________________________


### 3.Compilation step

* Three things you can change optimizer , loss and metrices 

In [27]:
neural_net.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])

### 4.Data preperation

* since we are not doing any type of convolution or pooling on the dataset ,than we have to manually flatten the images and passes it to the fully connected layers.


In [13]:
train_img = train_images.reshape((60000, 28 * 28))
train_img = train_img.astype('float32') / 255

test_img = test_images.reshape((10000, 28 * 28))
test_img = test_img.astype('float32') / 255

### 5.Label preparation

* labels must be one-hot encoded .

In [14]:
from keras.utils import to_categorical 

In [15]:
train_cls=to_categorical(train_labels)
test_cls=to_categorical(test_labels)

###  6.Fit the model to the training set.

In [28]:
neural_net.fit(train_img, train_cls, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 5s 86us/step - loss: 0.2304 - acc: 0.9335
Epoch 2/5
60000/60000 [==============================] - 6s 96us/step - loss: 0.0905 - acc: 0.9718
Epoch 3/5
60000/60000 [==============================] - 6s 94us/step - loss: 0.0602 - acc: 0.9802
Epoch 4/5
60000/60000 [==============================] - 6s 93us/step - loss: 0.0436 - acc: 0.9863
Epoch 5/5
60000/60000 [==============================] - 6s 97us/step - loss: 0.0342 - acc: 0.9889


* The training accuracy is 98.9% .


### 7.Evaluating the test set.


In [34]:
test_loss,test_acc=neural_net.evaluate(test_img,test_cls)
print('The test accuracy is ', test_acc*100)

10000/10000 [==============================] - 0s 45us/step
The test accuracy is  97.96000000000001


#                      END